[MLIR][transform][python] add sugared python abstractions for transform dialect

martin-luecke opened the pull request:
This adds Python abstractions for the different handle types of the transform dialect 

The abstractions allow for straightforward chaining of transforms by calling their member functions.
As an initial PR for this infrastructure, only a single transform is included: `transform.structured.match`. 
With a future `tile` transform abstraction an example of the usage is: 
```Python
def script(module: OpHandle):
    module.match_ops(MatchInterfaceEnum.TilingInterface).tile(tile_sizes=[32,32])
```
to generate the following IR:
```mlir
%0 = transform.structured.match interface{TilingInterface} in %arg0
%tiled_op, %loops = transform.structured.tile_using_for %0 [32, 32]
```

These abstractions are intended to enhance the usability and flexibility of the transform dialect by providing an accessible interface that allows for easy assembly of complex transformation chains.

martin-luecke committed:
[MLIR][transform][python] add extras for sugared python abstractions

martin-luecke committed:
cleaning up comments

martin-luecke committed:
add in-depth documentation to `insert_transform_script`

martin-luecke committed:
remove unneeded import

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+from __future__ import annotations
```
License header please.

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+        return self._mlir_value
+
+
+@dataclass
+class Param(Value):
+    """Wrapper around a transform Param with methods to chain further transforms."""
+
+
```
This looks unused/untested, please either add a test or remove.

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+    """Wrapper around a transform value handle with methods to chain further transforms."""
+
+    _mlir_value: ir.Value
+    children: list[Value] = field(default_factory=list)
+    parent: Optional[Value] = None
```
Any way to document these fields?

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+    module {
+        transform.sequence failures(propagate) {
+        ^bb0(%arg0: !transform.any_op):
+            %0 = transform.structured.match ops{["scf.for"]} in %arg0 : (!transform.any_op) -> !transform.op<"scf.for">
```
Nit: let's break the line somewhere to make it shorter.

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+
+    with module.context, ir.Location.unknown(module.context):
+        with ir.InsertionPoint.at_block_begin(module.body):
+            sequence_op = transform.SequenceOp(
```
Most of the upstream code has transitioned to using `NamedSequence` instead and we will be moving away from allowing a top-level sequence.

martin-luecke replied:
This now accepts a `Block` or `InsertionPoint` and creates a `NamedSequence` instead. 
However, the user currently has to make sure to set the attribute 'transform.with_named_sequence' on the parent symbol table, i.e. the module operation in most cases.
If we expose a `get_parent_of_type` (or something to get the parent symbol table) to Python at some point, we could add this automatically.

ftynse commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def insert_transform_script(
+    module: ir.Module,
```
Nit: does this have to be a module?

ftynse reviewed (approved):
This is consistent with other recently added usability extras. LGTM when transitioned to using named sequences. Please give Maks a change to review.

makslevental commented on mlir/python/CMakeLists.txt:
```diff
@@ -311,6 +311,14 @@ declare_mlir_dialect_python_bindings(
+  ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/mlir"
+  GEN_ENUM_BINDINGS
+  SOURCES
+    dialects/transform/extras/__init__.py)
```
umm can we keep all extras under one `extras`? one of these should be suitable:

1. `dialects/extras/transform/__init__.py`
2. `extras/dialects/transform/__init__.py`

The second is preferrable because moving `extras/types.py` to ~ `dialects/extras/types.py` doesn't really make sense.

martin-luecke replied:
Thank you for the suggestion, I went with 2. 👍

makslevental commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+                transform.AnyOpType.get(),
+            )
+        with ir.InsertionPoint(sequence_op.body):
+            script(OpHandle(sequence_op.bodyTarget))
```
You don't need to do this - you can [`register_value_caster(OpHandle)`](https://github.com/llvm/llvm-project/blob/7c850867b9ef4427375da6d83c34d0b9c944fcb8/mlir/test/python/ir/value.py#L372) and it'll automagically be done for you _everywhere_ `ir.Values` are returned from the bindings. Doing this will have the added benefit that you don't need the proxy/shadow values - you're "casted" values will just be subclasses of `ir.Value` already (i.e., you'll be able to ditch the indirection through `self.mlir_value`).

Quickly reminding myself about how the transform dialect bindings work, the only thing you're missing there is exposure of `TypeID` for some of the transform types (looks like [`OperationTypeGetTypeID`](https://github.com/llvm/llvm-project/blob/97f9f1a08ab1f5f91282cf95d13f306d03dc0888/mlir/lib/CAPI/Dialect/Transform.cpp#L64) is exposed but none of the others).

martin-luecke replied:
I went with your suggestion and exposed the corresponding typeid getters

makslevental commented on mlir/python/mlir/dialects/transform/extras/__init__.py:
```diff
@@ -0,0 +1,126 @@
+    from .... import ir
+    from ....dialects import transform
+    from ....dialects.transform import structured
+except ImportError as e:
+    raise RuntimeError("Error loading imports") from e
```
Nit: I don't think you really need this?

makslevental commented:
@ftynse @martin-luecke 

I asked for changes but (as I've mentioned before), I'm not opinionated on them - I think the changes I proposed are good but maybe you guys have already considered and discounted them - so if you guys don't agree I don't have any problem approving.

martin-luecke committed:
Address PR comments
- expose `get_static_typeid` for Python mlir type subclasses
- expose typeid getters for transform dialect types to CAPI
- sequence -> named_sequence
- more general `insert_transform_script`
- use `value_caster` automation instead of composition
- clarifying documentation
- cleaning up imports
- add missing license header

martin-luecke commented on mlir/include/mlir/Bindings/Python/PybindAdaptors.h:
```diff
@@ -495,6 +495,8 @@ class mlir_type_subclass : public pure_subclass {
           .attr("replace")(superCls.attr("__name__"), captureTypeName);
     });
     if (getTypeIDFunction) {
+      def_staticmethod("get_static_typeid",
+                       [getTypeIDFunction]() { return getTypeIDFunction(); });
```
This should probably better be `def_property_readonly_static(...` as defined in Pybind and used by `PyConcreteType` [here](https://github.com/llvm/llvm-project/blob/fd1ddd1bca4d43c49862274c12981b36e7a1ffe8/mlir/lib/Bindings/Python/IRModule.h#L950), but I am missing context to reimplement this.

makslevental replied:
Ah now I remember why I didn't forward - because I couldn't get all the way to implementing `def_property_readonly_static` up above. Lemme try again.

makslevental replied:
@martin-luecke can you try this patch

```diff
diff --git a/mlir/include/mlir/Bindings/Python/PybindAdaptors.h b/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
index 125f9b92937b..e0fd338f7109 100644
--- a/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
+++ b/mlir/include/mlir/Bindings/Python/PybindAdaptors.h
@@ -18,6 +18,7 @@
 #ifndef MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 #define MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 
+#include <pybind11/detail/internals.h>
 #include <pybind11/pybind11.h>
 #include <pybind11/pytypes.h>
 #include <pybind11/stl.h>
@@ -357,6 +358,17 @@ public:
     return *this;
   }
 
+  template <typename Func, typename... Extra>
+  pure_subclass &def_property_readonly_static(const char *name, Func &&f,
+                                              const Extra &...extra) {
+    py::cpp_function cf(std::forward<Func>(f), py::name(name), extra...);
+    auto *staticPropertyType = py::detail::get_internals().static_property_type;
+    auto builtinProperty =
+        py::reinterpret_borrow<py::object>((PyObject *)staticPropertyType);
+    thisClass.attr(name) = builtinProperty(cf);
+    return *this;
+  }
+
   template <typename Func, typename... Extra>
   pure_subclass &def_classmethod(const char *name, Func &&f,
                                  const Extra &...extra) {
@@ -495,8 +507,9 @@ public:
           .attr("replace")(superCls.attr("__name__"), captureTypeName);
     });
     if (getTypeIDFunction) {
-      def_staticmethod("get_static_typeid",
-                       [getTypeIDFunction]() { return getTypeIDFunction(); });
+      def_property_readonly_static("get_static_typeid", [getTypeIDFunction]() {
+        return getTypeIDFunction();
+      });
       py::module::import(MAKE_MLIR_PYTHON_QUALNAME("ir"))
           .attr(MLIR_PYTHON_CAPI_TYPE_CASTER_REGISTER_ATTR)(
               getTypeIDFunction())(pybind11::cpp_function(
```

martin-luecke replied:
This approach has quite a rats tail attached to it because it seems to change the behavior of how other fields are handled. This leads to lots of errors like `AttributeError: 'pybind11_static_property' object attribute '__doc__' is read-only`. 
Pybind uses [this code here](https://github.com/pybind/pybind11/blob/7d538a42750c8580eeaac10e505840a3694b04c8/include/pybind11/detail/class.h#L539) to work around this behavior. 
I will for now leave this as a static function to not introduce more complexity to `pure_subclass` just for this.

martin-luecke committed:
clarify documentation

martin-luecke committed:
add comment

makslevental commented on mlir/python/mlir/extras/dialects/transform/__init__.py:
```diff
@@ -0,0 +1,148 @@
+            isinstance(ops, str) and ops in structured.MatchInterfaceEnum.__members__
+        ):
+            if isinstance(ops, str):
+                ops = structured.MatchInterfaceEnum[ops]
```
Hmm makes me think we should change the emitted enums to handle strings as well like this (not in this PR obv).

makslevental commented on mlir/python/mlir/extras/dialects/transform/__init__.py:
```diff
@@ -0,0 +1,148 @@
+
+    def match_ops(
+        self,
+        ops: str
+        | ir.OpView
+        | structured.MatchInterfaceEnum
+        | Sequence[str | ir.OpView],
```
[This breaks compat with < 3.10](https://peps.python.org/pep-0604/). Personally I'd be happy to bump to 3.10 but prior we've tried to maintain down to 3.8 (I guess buildbots have upgraded because I've failed tests before with this).

ftynse replied:
The overall guidance is to have LLVM build on live LTS repos so it can be distributed. Ubuntu 20.04 is one of these, and it has python 3.8.2. Debian is surprisingly fresher, and I'm not aware what RHEL-based distros folks run these days. But sticking to 3.8 sounds reasonable. I don't know if we can import from future for this.

martin-luecke replied:
The future import is available since 3.7 and the `from __future__ import annotations` import is exactly why this does not break the build in Python 3.8. 
With this import the annotations here are [not evaluated by the Python interpreter](https://peps.python.org/pep-0563/) at module import time, but only when e.g. `typing.get_type_hints` is called on it (which we do not). The annotations are still understood by tools like MyPy, see [here](https://mypy.readthedocs.io/en/latest/runtime_troubles.html#using-x-y-syntax-for-unions). 
This should not be an issue for us, but I can still move this over to the `Union` syntax if you think this prudent.

makslevental replied:
I didn't know this! Definitely keep the pipes!

stellaraccident replied:
Our policy to date has been to track the PSF's eol policy, not specific LTS. In any case, it is a distinction without a difference in this case: we support 3.8 until it is eol in Oct-2024.

For the older Python versions, I am fine with conditional coding or some level of heroic if the feature is very useful, but in most cases I see, it is just easier to write it in an older-version compatible way and not deal with the mental burden of conditional stuff.

Feel free to plus me in on anything that needs untangling in this vein. Happy to offer advice.

stellaraccident commented on mlir/python/mlir/extras/dialects/transform/__init__.py:
```diff
@@ -0,0 +1,148 @@
+#  See https://llvm.org/LICENSE.txt for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from __future__ import annotations
```
Taking this from another patch where I saw this being added: I would very much prefer that transform dialect stuff be contained under its own dialect namespace, not in a parallel dialect tree. If it is experimental and needs more bake time, note that in the name or something, but let's not sprawl dialect specific things into other parts of the codebase.
